Create plugins flux store in redux

ryelle opened the pull request:
This adds a new subtree to our redux state for `installed`, which tracks installed plugins on sites. This also adds the selectors & actions to install, update, activate, deactivate, remove, and toggle autoupdates. See `client/state/plugins/installed/README.md` for a full list of actions, selectors, and an example of the state subtree.

Actions, reducer, and selectors _should_ all have full test coverage, but I might be missing edge cases. **To test:**

`npm run test-client client/state/plugins/installed`

This is not implemented in any plugin components in this PR, the implementation & removal of the flux stores will happen over a series of PRs by folks from team luna.

ryelle commented:
cc @johnHackworth @roccotripaldi @oskosk @tyxla - this is the first step (adding the redux store) I mentioned on our team chat. Pinging everyone since I don't remember who was interested :)

johnHackworth commented:
woooooooow ... this looks great. let's also ping  @lezama & @enejb

johnHackworth commented:
Well, this PR is huge ... but I don't see anything that doesn't look right. I would appreciate more reviews, specially from other people with experience with this part of the codebase ( *cough* @enejb *cough* ), but I think the best approach would be to merge it (since it's still not being used from anywhere) and once we migrate the components to make use of it, make sure that we are not breaking anything in the process

blowery commented on client/state/plugins/installed/actions.js:
```diff
@@ -0,0 +1,370 @@
+	return pluginHandler;
+};
+
+const getSolvedPromise = ( dataToPass ) => {
```
You could use `Promise.resolve` and `Promise.reject` instead of writing them yourself.

blowery commented on client/state/plugins/installed/actions.js:
```diff
@@ -0,0 +1,370 @@
+};
+
+export default {
+	activate: function( site, plugin ) {
```
We've been preferring named exports instead of exporting an object with properties. IE `export function activate(...)`. It makes the import side a bit more predictable when using `require`.

tyxla commented:
All of this looks awesome! 

One minor thing - I think that in addition to `isRequesting` and `hasRequested` selectors we could have `isRequestingForSites` and `hasRequestedForSites`, which could work with an array of sites. These could be useful for the `QueryPlugins` query component. Unless I'm misunderstanding the purpose of that query component. @ryelle thoughts?

ryelle commented:
Sure, requesting selectors for an array of sites would work well with `getPlugins`, since it also accepts an array of sites. I'll add that now 👍

enejb commented:
Thanks for taking this on @ryelle ❤️. 
A few things that I noticed.
- currently when we enable autoupdates when if the plugin needs an update we also do that right away See https://github.com/Automattic/wp-calypso/blob/master/client/lib/plugins/actions.js#L533
- There is also currently an action that gets fired silently which updates plugins right after plugins load see https://github.com/Automattic/wp-calypso/blob/master/client/lib/plugins/actions.js#L206 
this updates plugins for sites that have their sites not autoupdate plugins because of version control. 
- All actions are queued up ( so that they don't happen right away but one after another) since that resulted in actions failing. That is the reason behind https://github.com/Automattic/wp-calypso/blob/master/client/lib/plugins/actions.js#L21
I don't think I saw anything like that in the code (maybe I missed it) 

I ❤️  the documentation in the readme. I think it is missing the new selectors that you just added.

I ran the tests and beside the 3 points I made above I didn't find anything that stood out. 

Thanks again for your great work!

ryelle commented:
I was initially trying to decouple the actions from each other as much as possible, so I left those out. But then the process needed for installing also does updates, activation, autoupdates, etc... So yes, we should bring the updates actions back into autoupdates & the site plugins fetch.

> All actions are queued up

Is that because multiple plugin actions could be triggered on a single site? You're right, there's not any protection against that in the current code. I wonder if there is a good redux-y way to do this?

>  I think it is missing the new selectors that you just added.

It is, thanks for catching that 😄

enejb commented:
> Is that because multiple plugin actions could be triggered on a single site? 
Yes, multiple actions fire-ing before others are completed already don't work to tell on the single site. 
Sorry I don't know of a good redux-y way of doing this. Maybe @aduth has a suggestion?

aduth commented on client/state/plugins/installed/actions.js:
```diff
@@ -0,0 +1,360 @@
+	return pluginHandler;
+};
+
+export function activate( site, plugin ) {
```
It'd be nice if these action creators could simply accept `siteId`, since that's a bit easier to attain and work with (avoids potential for calling property on undefined in some cases where the action creator is called incorrectly).

aduth replied:
I think the naming of the action creators should be more specific to the context to avoid namespace conflicts if either several action creators are imported within the same file (without needing to rename them during import) and if we were ever to do something like [global selectors](https://github.com/Automattic/wp-calypso/pull/5954) for action creators.

aduth commented on client/state/plugins/installed/actions.js:
```diff
@@ -0,0 +1,360 @@
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_UPDATE_REQUEST_FAILURE, error } ) );
+		};
+
+		if ( ! site.canUpdateFiles ) {
```
To earlier point on passing `site` object, to workaround it here, we'd typically not have validation in an action creator but instead at the point where the action creator would be dispatched. The boilerplate is usually avoided by consolidating it within a query component, though it's a little less obvious how we'd handle that for update actions like this.

aduth commented on client/state/plugins/installed/actions.js:
```diff
@@ -0,0 +1,360 @@
+
+export function enableAutoupdate( site, plugin ) {
+	return ( dispatch ) => {
+		if ( ! utils.userCan( 'manage_options', site ) || ! site.canAutoupdateFiles ) {
```
It'd be nice if this were validated using the existing [`canCurrentUser` selector](https://github.com/Automattic/wp-calypso/blob/cefb6aa92d6f68b5dcbde0c92be7b5bc5aff0211/client/state/current-user/selectors.js#L91-L110) outside of the action creator.

aduth commented on client/state/plugins/installed/actions.js:
```diff
@@ -0,0 +1,360 @@
+
+		const successCallback = ( data ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_AUTOUPDATE_ENABLE_REQUEST_SUCCESS, data } ) );
+			this.update( site, plugin );
```
What is this calling? Should this be `dispatch( update( site, plugin ) );` ?

aduth commented on client/state/plugins/installed/reducer.js:
```diff
@@ -0,0 +1,146 @@
+				...state.slice( 0, index ),
+				...state.slice( index + 1 )
+			];
+		default:
```
If you're intending to persist state, you should also provide a schema to validate against ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-persistence--2754-)).

You may also be interested in the [`createReducer` utility](https://github.com/Automattic/wp-calypso/blob/cefb6aa92d6f68b5dcbde0c92be7b5bc5aff0211/client/state/utils.js#L44-L118) which helps avoid most of the boilerplate (reminds me it needs to be documented better 😄 )

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+};
+
+export default {
+	isRequesting,
```
Instead of exporting these as a single default export object, why not just export them as named exports at the point they're declared? This could potentially have an effect on future [tree-shaking](http://www.2ality.com/2015/12/webpack-tree-shaking.html) too.

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+
+const isRequesting = function( state, siteId ) {
+	// if the `isRequesting` attribute doesn't exist yet,
+	// we assume we are still launching the fetch action, so it's true
```
This assumption seems strange to me. Why would we return true for `isRequesting` if the request hasn't started?

tyxla replied:
I agree with @aduth on this. I think we should return `false` in that case, because the request hasn't started. 

If you are still in doubt, have a look at the `QueryPlugins` component (it's still in progress, but anyway). If you look [here](https://github.com/Automattic/wp-calypso/blob/1ed9ca66f54a926f70f157d972b71ccc20e0043e/client/components/data/query-plugins/index.jsx#L17) you will notice that the initial fetch will never occur, since `this.props.isRequestingForSites` is initially `true`.

ryelle replied:
yeah, it's already changed locally, I just didn't push last night since I'm still working through other feedback.

aduth commented on client/state/plugins/installed/reducer.js:
```diff
@@ -0,0 +1,146 @@
+/*
+ * Tracks the requesting state for installed plugins on a per-site index.
+ */
+export function hasRequested( state = {}, action ) {
```
What will this be used for? To avoid re-requesting plugins for a site if they've already been received? Typically we'd be liberal in allowing a query component to re-request in its lifecycle methods regardless of whether it'd been requested in the past.

ryelle replied:
It can be removed, in that case -- I don't think we'd need to be strict about it. I was following another reducer that did need to know.

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+
+const getPluginOnSite = function( state, site, pluginSlug ) {
+	const pluginList = getPlugins( state, [ site ] );
+	return find( pluginList, { slug: pluginSlug } ) || false;
```
Why is `false` the fallback value instead of `null` or `undefined`?

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+	return every( sites, ( siteId ) => hasRequested( state, siteId ) );
+};
+
+const getPlugins = function( state, sites, pluginFilter = false ) {
```
Why do we need a default value for `pluginFilter`? Would appear to work just as well without.

aduth replied:
This appears to be a heavier selector. You might consider using [the `createSelector` utility](https://github.com/Automattic/wp-calypso/tree/master/client/lib/create-selector) to cache its result.

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+			}
+		} );
+	} );
+	if ( !! pluginFilter && _filters[ pluginFilter ] ) {
```
I don't see any need to coerce to a boolean here. `if ( pluginFilter &&` should work just as well

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+	let pluginList = {};
+	forEach( sites, ( site ) => {
+		const list = state.plugins.installed.plugins[ site.ID ] || [];
+		list.map( ( item ) => {
```
This doesn't appear to be an appropriate use of [`Array#map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) since you're not using the mapped result. `Array#forEach` is probably more applicable.

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+
+const getPlugins = function( state, sites, pluginFilter = false ) {
+	let pluginList = {};
+	forEach( sites, ( site ) => {
```
The way you're generating the value for `pluginList` from the `forEach` makes me think it'd be a good use case for `Array#reduce`:

```js
let pluginList = sites.reduce( ( memo, site ) => /* ... */, {} );
```

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+
+const getPluginsWithUpdates = function( state, sites ) {
+	const pluginList = getPlugins( state, sites );
+	if ( pluginList.length === 0 ) {
```
`filter` will handle the empty array fine. I assume it'd be negligible in terms of performance difference and would clean up the selector a bit. Could even get it down to a one-liner:

```js
return filter( getPlugins( state, sites ), _filters.updates );
```

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+		list.map( ( item ) => {
+			// plugin has a sites property which lists the sites this plugin is installed on
+			// each site, in turn, has a plugin property for the specific plugin's info.
+			const siteWithPlugin = Object.assign( {}, site, { plugin: item } );
```
I'm not sure how it'll end up being used, but we should be careful if we were to rely on the full site object reference within this selector's result. I worry we might run into some syncing issues, and it'd be better if we could trust `state.sites` to be the source of truth instead (i.e. store IDs and force lookup via `getSite` selector).

aduth commented on client/state/plugins/installed/selectors.js:
```diff
@@ -0,0 +1,164 @@
+		return [];
+	}
+
+	const pluginSites = uniq( compact(
```
What is `uniq` doing here? Seems like we shouldn't have duplicate sites within `plugin.sites` and also not sure how well it'd be able to find uniques, since they're objects.

aduth commented on client/state/plugins/installed/status/reducer.js:
```diff
@@ -0,0 +1,135 @@
+					[ siteId ]: statusForSite( state[ siteId ], action )
+				} );
+			}
+			return Object.assign( {}, state, { [ siteId ]: statusForSite( {}, action ) } );
```
Couldn't this case block be simplified to:

```js
Object.assign( {}, state, { [ siteId ]: statusForSite( state[ siteId ], action ) } );
```

Since even if `state.hasOwnProperty( siteId )` is `false`, the `statusForSite` default argument for `state` will provide the empty object.

aduth commented on client/state/plugins/installed/test/actions.js:
```diff
@@ -0,0 +1,554 @@
+		spy.reset();
+	} );
+
+	after( () => {
```
The [`useNock` test helper](https://github.com/Automattic/wp-calypso/tree/master/test/test/helpers/use-nock) can help eliminate some of the boilerplate. See #7875 as an example of the changes necessary.

aduth commented on client/state/plugins/installed/test/actions.js:
```diff
@@ -0,0 +1,554 @@
+	DISABLE_AUTOUPDATE_PLUGIN
+} from '../constants';
+import {
+	fetch as fetchPlugins,
```
Another point to why it'd be nice if the renamed import here was simply the original name of the action creator 😄

aduth commented on client/state/plugins/installed/test/plugins.js:
```diff
@@ -0,0 +1,72 @@
+/* eslint-disable max-len */
```
If you have test data used for mocks, they should be in a subdirectory within `test` because of the way test discovery works (i.e. as-is, Mocha will try to process this file as a test file).

See: https://github.com/Automattic/wp-calypso/tree/master/test#how-to-add-a-new-test-file

aduth commented on client/state/plugins/installed/test/reducer.js:
```diff
@@ -0,0 +1,239 @@
+/* eslint-disable max-len */
```
Do you have a good reason for disabling `max-len` in this file? The long lines I see could easily be refactored onto separate lines.

aduth commented on client/state/plugins/installed/test/reducer.js:
```diff
@@ -0,0 +1,239 @@
+ */
+import { expect } from 'chai';
+import deepFreeze from 'deep-freeze';
+import omit from 'lodash/omit';
```
Since #6539 you can now import directly from lodash:

```js
import { omit } from 'lodash';
```

aduth commented on client/state/plugins/installed/test/selectors.js:
```diff
@@ -0,0 +1,330 @@
+/* eslint-disable max-len */
```
Same point about `max-len` as earlier.

aduth commented:
>Sorry I don't know of a good redux-y way of doing this. Maybe @aduth has a suggestion?

Could the queue be tracked as a reducer? Removing an item when from the queue it finishes and continuing to pop items off to dispatch as long as there is anything in the queue?

ryelle commented on client/state/plugins/installed/actions.js:
```diff
@@ -0,0 +1,360 @@
+	return ( dispatch ) => {
+		const pluginId = getPluginId( site, plugin );
+		const defaultAction = {
+			action: ACTIVATE_PLUGIN,
```
It's saved into the status reducer, so that we know from `getStatusForPlugin` what action is happening (activating, installing, etc)

ryelle commented:
I've addressed _most_ of the feedback here (and rebased last week for the node update). For organization, what hasn't happened yet:

- No `createSelector` yet, the selector & tests need to be updated to accept just siteIds
- After fetching plugins, we want to be able to trigger an update for all sites' plugins
- The queue for plugin actions is not in place yet

johnHackworth commented:
Since this is still not being used by anything in production, it's huge and it's kind of blocking further developments that depends on this redux subtree, I think we can merge it like it is right now, and then work in the rest of the still existing issues in another PR. I'm going to mark this as ready to merge and we'll start working on everything else right away, in another branch
